server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,168 @@
+
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                parseTypeSignature("array(E)"),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.BOOLEAN.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature()
+            ).withTypeVariableConstraints(typeVariable("E")),
+            GenerateSubscripts::new
+        );
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private GenerateSubscripts(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    private static List<?> extractNestedArray(List<?> array, int targetLevel) {
+        List<?> nested = array;
+        for (int level = 2; level <= targetLevel; level++) {
+            if (nested == null || nested.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || e instanceof List)) {
+                throw new IllegalArgumentException("expected a nested array at level " + level);
+            }
+            List<?> head = (List<?>) nested.get(0);
+            if (head == null || head.isEmpty()) {
+                return null;
+            }
+            if (!nested.stream().allMatch(e -> e == null || ((List<?>) e).size() == head.size())) {
+                throw new IllegalArgumentException("nested arrays should have the same dimension within a level");
+            }
+            nested = head;
+        }
+        return nested;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+
+        List<?> array = (List<?>) args[0].value();
+        if (array == null || array.isEmpty()) {
+            return () -> NO_RESULTS_ITERATOR;
+        }
+        int dim = (Integer) args[1].value();
+        if (dim <= 0) {
+            throw new IllegalArgumentException("dimension must be greater than zero");
+        }
+        List<?> nestedArray = extractNestedArray(array, dim);
+        if (nestedArray == null || nestedArray.isEmpty()) {
+            return () -> NO_RESULTS_ITERATOR;
+        }
+        boolean reversed = args.length == 3 && (Boolean) args[2].value();
+        return () -> new Iterator<>() {
+
+            final Object [] columns = new Object[]{ null };
+            final RowN row = new RowN(columns);
+            final Iterator<Integer> subscripts = IntStream
```

Comment:
I am not sure that stream usage is justified here. It generates the boxed sequence, sorts it if needed, allocates additional unnecessary objects, etc. As an idea, can't we just have two int fields, e.g. `from` and `to` that we set depending on the reverse flag, also probably we need the sign, so we know whether we inc or dec or smth similar? Something like that will be definitely more efficient than the current stream API usage.